common/web/types/src/util/util.ts

```diff
@@ -350,5 +353,36 @@ export class BadStringAnalyzer {
     }
   }
 
+  /** internal map */
   private m = new BadStringMap();
 }
+
+/** analyze a string looking for bad unicode */
+export class BadStringAnalyzer extends StringAnalyzer {
+  /** analyze one codepoint */
+  protected analyzeCodePoint(c: string, ch: number): BadStringType {
+    return BadStringAnalyzer.getProblem(ch);
+  }
+  /** export analyzer function  */
+  public static getProblem(ch: number) {
+    if (!isValidUnicode(ch)) {
+      return BadStringType.illegal;
+    } else if(isPUA(ch)) {
+      return BadStringType.pua;
+    } else { // TODO-LDML: unassigned
+      return null;
+    }
+  }
+}
+
+/** Analyzer that checks if something isn't NFD */
+export class NFDAnalyzer extends StringAnalyzer {
+  protected analyzeCodePoint(c: string, ch: number): BadStringType {
+    const nfd = c.normalize("NFD");
+    if (c !== nfd) {
+      return BadStringType.denormalized;
+    } else {
+      return null;
+    }
+  }
```

Comment:
I wonder about the performance of this -- seems like it could be quite expensive?

Reply:
right. maybe some kind of cache would help.

Reply:
yes, could perhaps cache?

Reply:
it's only on actual ranges in transforms.

Reply:
OK, limited use so we can probably ignore for now. If it's always a single codepoint we could build a map and ship it -- trivial. Danger with that is that if we use `normalize()` elsewhere then the map could be inconsistent wrt Unicode version.